tests/cpu_bugs/xen_mitigations.pm

```diff
@@ -0,0 +1,349 @@
+            }
+        },
+        "l1d-flush=off" => {
+            "default" => {
+                expection   => {''},
+                unexpection => {'xl dmesg' => ['Xen settings: BTI-Thunk JMP, SPEC_CTRL: IBRS+ SSBD-, Other: .*L1D_FLUSH']},
+            }
+        },
+        "l1d-flush=on" => {
+            "default" => {
+                expection   => {'xl dmesg' => ['Xen settings: BTI-Thunk JMP, SPEC_CTRL: IBRS+ SSBD-, Other: .*L1D_FLUSH']},
+                unexpection => {''},
+            }
+        },
+    },
+  };
+sub do_check {
+    my $secnario = shift;
+    my $foo      = $secnario->{'default'};
+    if ($foo->{'expection'}) {
+        while (my ($cmd, $lines) = each %{$foo->{'expection'}}) {
+            foreach my $l (@{$lines}) {
+                if ($l ne "") {
+                    assert_script_run("$cmd | grep \"$l\"");
+                } else {
+                    print "This expection is empty string, skip";
+                }
+
+            }
+        }
+    }
+    if ($foo->{'unexpection'}) {
+        while (my ($cmd, $lines) = each %{$foo->{'unexpection'}}) {
+            foreach my $l (@{$lines}) {
+                if ($l ne "") {
+                    my $ret = script_run("$cmd | grep \"$l\"");
+                    if ($ret eq 0) {
+                        die "found a unexpection string.";
+                    }
+                } else {
+                    #Debug what output be report.
+                    assert_script_run("xl dmesg | grep -A 10 \"Speculative\"");
+                    print "This unexpection is empty string, skip";
+
+                }
+
+            }
+        }
+    }
+}
+
+sub do_test {
+    my ($self, $hash) = @_;
+    #xen parameter be store into arg
+    #
+    while (my ($arg, $dict) = each %$hash) {
+        while (my ($key, $value) = each %$dict) {
+            my $parameter = $arg . '=' . $key;
+            bootloader_setup::add_grub_xen_cmdline_settings($parameter);
+            bootloader_setup::grub_mkconfig();
+            Mitigation::reboot_and_wait($self, 150);
+            #$value include the check rules of current $parameter.
+            my $ret = do_check($value);
+            if ($ret ne 0) {
+                record_info('ERROR', "$parameter test is failed.", result => 'fail');
+                die;
+            }
+            record_info('INFO', "$parameter test is finished.");
+            bootloader_setup::remove_grub_xen_cmdline_settings($parameter);
+            bootloader_setup::grub_mkconfig();
+        }
+
+    }
+}
+
+
+sub run {
+    my $self = @_;
+    select_console 'root-console';
+    my $ret = script_run("test -d /proc/xen");
+    if ($ret ne 0) {
+        record_info("platform mistake, This system is not running as Dom0.");
+        die;
+    }
+    do_test($self, $mitigations_list);
+}
+
+
+sub post_fail_hook {
+    my ($self) = @_;
+    select_console 'root-console';
+    assert_script_run(
```

Comment:
generally not good idea to assert something in post fail hook , you already fail what this assertion will give you ? if you really want to fail here let's move this command after grub changes . Don't think that upload logs any how related to the fact that you need to change grub . 
BTW don't you need to do the same if test pass ?

Reply:
script_run is enough, I think, just collect logs.